### PR TITLE
Flash boards with JLink firmware using JlinkExe

### PR DESCRIPTION
**Pull request by juhhov**
## Status
**READY**

## Description
Drag and drop method is not robust with the current JLink firmware.
Use JLinkExe to flash JLink boards.

[Jira](https://jira.arm.com/browse/IOTSYSTOOL-647)

## Known issues
JLinkExe does not handle unsuitable binary in any way.

## Todo
- [x] Tests
- [x] Documentation

### CODE REVIEW

File: mbed_flasher/flashers/FlasherJLink.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+            self.logger.info("Flash of {} succeeded".format(target["target_id"]))
+            self.logger.debug(output)
+            return EXIT_CODE_SUCCESS
+        except subprocess.CalledProcessError as error:
+            self.logger.error("Flash of {} failed, with returncode {}"
+                              .format(target["target_id"], error.returncode))
+            self.logger.error(error.output)
+            return EXIT_CODE_FLASH_FAILED
+        finally:
+            cmd_script.close()
```
Comment (jonikula) [inline mbed_flasher/flashers/FlasherJLink.py]: Is there some reason this is not closed earlier?

Comment (juhhov) [reply mbed_flasher/flashers/FlasherJLink.py]: From NamedTemporaryFile docs:
 If delete is true (the default), the file is deleted as soon as it is closed.

File: mbed_flasher/flashers/__init__.py [hunk 2]
```diff
@@ -15,13 +15,17 @@
 limitations under the License.
 """
 
-from mbed_flasher.flashers.FlasherMbed import FlasherMbed as mbed_flasher
+from mbed_flasher.flashers.FlasherMbed import FlasherMbed
+from mbed_flasher.flashers.FlasherJLink import FlasherJLink
 
 # disable Invalid constant name warning, not a const
 # pylint: disable=C0103
-AvailableFlashers = [
```
Comment (YifeiZuo01) [inline mbed_flasher/flashers/__init__.py]: I remember this namespace changes seems were for some tests,  maybe here will have some errors.

File: mbed_flasher/flashers/FlasherJLink.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        FlasherBase.__init__(self, logger)
+
+    @staticmethod
+    def get_supported_targets():
+        """
+        :return: supported JLink types
+        """
+        if not FlasherJLink.supported_targets:
+            mbeds = mbed_lstools.create()
+            # @todo this is workaround until mbed-ls provide public api
```
Comment (jupe) [inline mbed_flasher/flashers/FlasherJLink.py]: is there task for mbedls for this ?

Comment (juhhov) [reply mbed_flasher/flashers/FlasherJLink.py]: Did you make one when you changed it?

Comment (jupe) [reply mbed_flasher/flashers/FlasherJLink.py]: ah this is same one? then there should be - or is there.. hmm. Need to check and raise if not exists..

File: test/test_flasherbase.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        self.assertTrue(hasattr(FlasherBase, 'supported_targets'))
+
+        self.assertTrue(hasattr(base, '_start_and_wait_flash'))
+        self.assertTrue(hasattr(base, '_flash_run'))
+
+    @mock.patch("mbed_flasher.flash.Logger")
+    def test_start_and_wait_flash_executes(self, logger):
+        base = FlasherBase(logger)
+        command = """
+import sys
```
Comment (jonikula) [inline test/test_flasherbase.py]: Why this weird formatting?

Comment (juhhov) [reply test/test_flasherbase.py]: No extra characters in the command which goes to command line

**Pull request merged**